src/producer.rs

```diff
@@ -87,6 +88,59 @@ pub fn to_substrait_rel(plan: &LogicalPlan) -> Result<Box<Rel>> {
                 }))),
             }))
         }
+        LogicalPlan::Join(join) => {
+            let left = to_substrait_rel(join.left.as_ref())?;
+            let right = to_substrait_rel(join.right.as_ref())?;
+            let join_type = match join.join_type {
+                JoinType::Inner => 1,
+                JoinType::Left => 2,
+                JoinType::Right => 3,
+                JoinType::Full => 4,
+                JoinType::Anti => 5,
+                JoinType::Semi => 6,
+            };
+            // we only support basic joins so return an error for anything not yet supported
+            if join.null_equals_null {
+                return Err(DataFusionError::NotImplemented(
+                    "join null_equals_null".to_string(),
+                ));
+            }
+            if join.filter.is_some() {
+                return Err(DataFusionError::NotImplemented("join filter".to_string()));
+            }
+            match join.join_constraint {
+                JoinConstraint::On => {}
+                _ => {
+                    return Err(DataFusionError::NotImplemented(
+                        "join constraint".to_string(),
+                    ))
+                }
+            }
+            // map the left and right columns to binary expressions in the form `l = r`
+            let join_expression: Vec<Expr> = join
+                .on
+                .iter()
+                .map(|(l, r)| Expr::Column(l.clone()).eq(Expr::Column(r.clone())))
+                .collect();
+            // build a single expression for the ON condition, such as `l.a = r.a AND l.b = r.b`
+            let join_expression: Expr = join_expression
+                .iter()
+                .skip(1)
```

Comment:
You might be able to use use into_iter with `reduce` here?

Reply:
Thanks @Dandandan. That is cleaner.